rosplan_sensing_interface/scripts/sensing_interface.py

```diff
@@ -0,0 +1,440 @@
+
+        ######
+        # Subscribe to all the topics
+        self.offset = {}  # Offset for reading cfg
+        for predicate_name, predicate_info in self.cfg_topics.iteritems():
+            params_loaded = self.load_params(predicate_name, predicate_info)  # If parameters found, add 1 to the indexes
+            if not params_loaded[0]:
+                continue
+            offset = self.offset[predicate_name] = int(params_loaded[1])
+            if len(predicate_info) < 2+offset:
+                rospy.logerr("Error: Wrong configuration file for predicate %s" % predicate_name)
+                continue
+            msg_type = predicate_info[offset+1]
+            self.import_msg(msg_type)
+            rospy.Subscriber(predicate_info[offset+0], eval(msg_type[msg_type.rfind('/')+1:]), self.subs_callback, predicate_name)
+            # self.sensed_topics[predicate_name] = (None, False)
+            try:  # Update KB to inform about the sensed predicates
+                sensed_srv_req = SetNamedBoolRequest()
+                sensed_srv_req.name = predicate_name
+                sensed_srv_req.value = True
+                set_sensed_predicate_srv.call(sensed_srv_req)
+            except Exception as e:
+                rospy.logerr('KCL: (RosplanSensing) Could not update sensing information in Knowledge Base for proposition %s',
+                             predicate_name)
+            rospy.loginfo('KCL: (RosplanSensing) Predicate %s: Subscribed to topic %s of type %s', predicate_name,
+                          predicate_info[offset+0], msg_type)
+
+        ############
+        # Create clients for all the services
+        self.service_clients = []
+        self.service_names = []
+        self.service_type_names = []
+        self.service_predicate_names = []
+        self.last_called_time = []
+        self.time_between_calls = []
+        self.request_src = []
+        self.response_process_src = []
+        for predicate_name, predicate_info in self.cfg_service.iteritems():
+            params_loaded = self.load_params(predicate_name, predicate_info)  # If parameters found, add 1 to the indexes
+            if not params_loaded[0]:
+                continue
+            offset = self.offset[predicate_name] = int(params_loaded[1])
+            if len(predicate_info) < 2+offset:
+                rospy.logerr("Error: Wrong configuration file for predicate %s" % predicate_name)
+                continue
+            srv_type = predicate_info[offset+1]
+            srv_typename = self.import_srv(srv_type)
+            self.service_type_names.append(srv_typename)
+            self.service_clients.append(rospy.ServiceProxy(predicate_info[offset+0], eval(srv_typename)))
+            self.service_names.append(predicate_info[offset+0])
+            self.service_predicate_names.append(predicate_name)
+            self.last_called_time.append(rospy.Time(0))
+            self.time_between_calls.append(predicate_info[offset+2])
```

Comment:
If a map is used in the yaml (as described above) and parsed to a dictionary, then the use of offset should not be required. Additionally the order would not matter in the config file.

Reply:
I like the idea, will try now.